### ISSUE TITLE
vim9script: Can't get "items" key of quickfix list from ":def" function

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#6357]
**Describe the bug**

Trying to get the `items` key of a quickfix list from inside a `:def` function raises E715.

**To Reproduce**

Run this shell command:

    vim -Nu NONE -S <(cat <<'EOF'
        def Func()
            helpg grail
            let qfl = getqflist({'items': 0})
            let items = qfl.items
        enddef
        call Func()
    EOF
    )

It raises `E715`:

    Error detected while processing function Func:
    line    3:
    E715: Dictionary required

**Expected behavior**

No error is raised, and `items` contains the items of the quickfix list populated by `:helpg grail`.

**Environment**

 - Vim version: 8.2 Included patches: 1-1079
 - OS: Ubuntu 16.04.6 LTS
 - Terminal: XTerm(356)

**Additional context**

I think this is a bug because the error comes from this line:

    let items = qfl.items

Which suggests that `qfl` is not a dictionary.  But it *is* a dictionary:

    vim -Nu NONE -S <(cat <<'EOF'
        def Func()
            helpg grail
            let qfl = getqflist({'items': 0})
            echom type(qfl) == v:t_dict
        "   ^-------------------------^
        enddef
        call Func()
    EOF
    )

    (1 of 1): You found it, Arthur!^I^I^I^I*holy-grail* *:smile*
    v:true
    ^----^

---

Besides, the same function works as expected when defined with `:fu` instead of `:def`.

---

Finally, the same code works as expected if I replace `getqflist({'items': 0})` with its literal dictionary evaluation:

    vim -Nu NONE -S <(cat <<'EOF'
        def Func()
            let qfl = {'id': 1, 'items': [{'lnum': 37557, 'bufnr': 2, 'col': 7, 'pattern': '', 'valid': 1, 'vcol': 0, 'nr': 0, 'type': '', 'module': '', 'text': '|holy-grail|'}, {'lnum': 1124, 'bufnr': 3, 'col': 32, 'pattern': '', 'valid': 1, 'vcol': 0, 'nr': 0, 'type': '', 'module': '', 'text': 'You found it, Arthur!				*holy-grail* *:smile*'}]}
            let items = qfl.items
        enddef
        call Func()
    EOF
    )

**Comment by bfrg** [comment]
The return types of `getqflist()` (and `getloclist()`) probably need to be changed to `any`, since the return type of the function is `list<dict<any>>` when called _without_ arguments, but when called _with_ an argument, the return type should be `dict<any>`.

https://github.com/vim/vim/blob/6378b21d6dd38cc0f80aa6d31d747db6c287483b/src/evalfunc.c#L597

You can get a clearer error message when you explicitly specify the type of the variable. For example:
```vim
    def Test_works()
        let qfl: list<dict<any>> = getqflist()
    enddef

    def Test_fail()
        let qfl: dict<any> = getqflist({'items': 0})
    enddef
    defcompile
```
The second function will print the following error:
```
E1013: type mismatch, expected dict<any> but got list<dict<any>>
```

**Comment by bfrg** [comment]
@lacygoill Wait why did you close it? 
@brammool Isn't this a bug?

The return type of `getqflist({'items': 0})` is definitively not `list<dict<any>>`. It should be `dict<list<any>>`. But on the other hand `getqflist({'title': 1})` returns `dict<string>`. So it depends on the arguments passed to `getqflist()`.

**Comment by lacygoill** [comment]
Oh, I thought you meant there was a mistake in my code, sorry.

**Comment by chrisbra** [comment]
fixed by f151ad1c70825a91afb112e611db5c712e2656ef
